src/plan_expand_hypertable.c

```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
 	pg_unreachable();
 }
 
+/*
+ * Constify expressions of the following form in WHERE clause:
+ *
+ * column OP timestamptz - interval
+ * column OP timestamptz + interval
+ *
+ * Iff interval has no month component.
+ *
+ * Since the operators for timestamptz OP interval are marked
+ * as stable they will not be constified during planning a
+ * subset of these operations are immutable and can be safely
+ * constified during planning. We adjust those expressions here
+ * and transform them for those cases where it's safe to do so
+ * during planning.
+ */
+static OpExpr *
+constify_timestamptz_op_interval(PlannerInfo *root, OpExpr *constraint)
+{
+	Assert(op->args->length == 2);
+
+	Expr *left = linitial(constraint->args);
+	Expr *right = llast(constraint->args);
+	OpExpr *op;
+	bool var_on_left = false;
+	Interval *interval;
+	Const *c_ts, *c_int;
+	Datum constified;
+
+	if (IsA(left, Var) && IsA(right, OpExpr))
+	{
+		op = castNode(OpExpr, right);
+		var_on_left = true;
+	}
+	else if (IsA(left, OpExpr) && IsA(right, Var))
+	{
+		op = castNode(OpExpr, left);
+	}
+	else
+		return constraint;
+
+	Oid plus_oid = ts_get_operator("+", PG_CATALOG_NAMESPACE, TIMESTAMPTZOID, INTERVALOID);
+	Oid minus_oid = ts_get_operator("-", PG_CATALOG_NAMESPACE, TIMESTAMPTZOID, INTERVALOID);
+
+	if (op->opresulttype != TIMESTAMPTZOID || (op->opno != plus_oid && op->opno != minus_oid) ||
+		op->args->length != 2 || !IsA(linitial(op->args), Const) || !IsA(llast(op->args), Const))
+		return constraint;
+
+	c_ts = linitial_node(Const, op->args);
+	c_int = llast_node(Const, op->args);
+
+	if (c_ts->consttype != TIMESTAMPTZOID || c_int->consttype != INTERVALOID || c_ts->constisnull ||
+		c_int->constisnull)
+		return constraint;
+
+	interval = DatumGetIntervalP(c_int->constvalue);
+
+	/*
+	 * constification is only safe when the interval has no month component
+	 * because month length is variable
+	 */
+	if (interval->month != 0)
+		return constraint;
+
+	if (op->opno == plus_oid)
+		constified =
+			DirectFunctionCall2(timestamptz_pl_interval, c_ts->constvalue, c_int->constvalue);
+	else
+		constified =
+			DirectFunctionCall2(timestamptz_mi_interval, c_ts->constvalue, c_int->constvalue);
+
+	c_ts = copyObject(c_ts);
+	c_ts->constvalue = constified;
+
+	if (var_on_left)
+		right = (Expr *) c_ts;
+	else
+		left = (Expr *) c_ts;
```

Comment:
I don't think you need to make this distinction (of left and right part of expression) and can just build the expression as "TS + C". It makes the code a lot easier to follow.

* C - TS is not allowed.
* TS + C is allowed.
* C + TS is allowed and equivalent to TS + C (actually rewritten to this form).
* C1 + TS - C2 is allowed and equivalent to (TS + C1) - C2 (actually rewritten to this form).

Reply:
yes you do because it will be a different operator for other configurations
TIMESTAMPTZ + INTERVAL is different operator from INTERVAL + TIMESTAMPTZ